### PR TITLE
chore: parallelize travis

### PR DESCRIPTION
**Pull request by virkt25**
### Changes
- Install top level dependencies without bootstrapping for all Jobs
- Bootstrap modules as needed for the Jobs
- Run Tests by Default
- Lint (Code / Commit Message) happen only on Linux + Node 8
    - They happen in parallel now instead of waiting for the `test` stage to complete.

## Checklist

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,6 +41,7 @@
     "build:full": "npm run clean:lerna && npm install && npm test",
     "pretest": "npm run clean && npm run build",
     "test": "node packages/build/bin/run-nyc npm run mocha --scripts-prepend-node-path",
+    "test:ci": "npm run pretest && npm run test -- --ignore-scripts",
```
Comment (bajtos) [inline package.json]: I find this confusing - how is `test:ci` different from `test`, when they both run `pretest` (build) and `test`?

File: bin/travis.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    exit 0
+  else
+    echo "Testing @loopback/docs"
+    npm run verify:docs
+  fi
+
+# Mocha Tests
+else
+  echo "TASK => MOCHA TESTS"
+  lerna bootstrap
```
Comment (bajtos) [inline bin/travis.sh]: Maybe `npm run postinstall` for consistency with other places where we defer the work to npm scripts?

IIRC, npm needs about ~50ms to execute the script, therefore your current solution is faster from performance view.

I don't have a strong opinion about this line, feel free to choose either way.

File: package.json [hunk 3]
```diff
@@ -13,6 +13,7 @@
     "@commitlint/cli": "^6.1.3",
     "@commitlint/config-conventional": "^6.1.3",
     "@commitlint/config-lerna-scopes": "^6.1.3",
+    "@commitlint/travis-cli": "^6.1.3",
```
Comment (bajtos) [inline package.json]: +1 for using an existing npm package instead of maintaining our own linter script 👏

**Comment by raymondfeng** [comment]
Is the proposed optimization needed because `commit lint` and `verify docs` are two independent jobs? Would be it be better off if we make them conditional steps of the `npm test`?

**Comment by virkt25** [comment]
We were using build stages before which required the previous stage to complete before the jobs in the next one would be kicked off. This optimization makes all jobs run in parallel on Travis since they are all a part of the same stage. This speeds up the Travis checks significantly (saves ~50 - 60 seconds). 

The reason I don't run `commit lint` and `verify docs` as part of the `test` job is two fold ... 
1. we run tests on multiple platforms and node versions and these tests only need to take place once. 
2. It's easier to look at the job that failed to know there is a linting error / docs failure rather than having to scroll to the bottom of one long log file to identify the issue.

### Before
image


### After
image

**Comment by raymondfeng** [comment]
@virkt25 Thanks for the explanation. 

Two more issues:
- I'm seeing the warning below on Travis. Please fix it.

```
loopback-next@0.1.0 test /home/travis/build/strongloop/loopback-next
node packages/build/bin/run-nyc npm run mocha --scripts-prepend-node-path "--ignore-scripts"
npm WARN invalid config scripts-prepend-node-path="--ignore-scripts"
```

- `npm run lint` is still executed as part of `posttest`. It's redundant of the `code-lint` task.

File: package.json [hunk 4]
```diff
@@ -42,6 +43,8 @@
     "build:full": "npm run clean:lerna && npm install && npm run build",
     "pretest": "npm run clean && npm run build:current",
     "test": "node packages/build/bin/run-nyc npm run mocha --scripts-prepend-node-path",
+    "pretest:ci": "npm run pretest",
```
Comment (bajtos) [inline package.json]: When running on a CI server, I believe we always start with a clean checkout, therefore there is no need to run `npm run clean`?

I am proposing:

```
"pretest:ci": "npm run build",
```

Not a big deal, feel free to land this pull request as it is now.

Comment (virkt25) [reply package.json]: Switched to `npm run build:current`

**Comment by bajtos** [comment]
Looks like this change is improving the speed of AppVeyor builds from 2x 7mins to 2x 6mins 👍  Can't wait to see it landed.

**Pull request merged**